homeassistant/components/prometheus/__init__.py

```diff
@@ -442,8 +444,8 @@ def _unit_string(unit):
             return
 
         units = {
-            TEMP_CELSIUS: "c",
-            TEMP_FAHRENHEIT: "c",  # F should go into C metric
+            TEMP_CELSIUS: "celsius",
+            TEMP_FAHRENHEIT: "celsius",  # F should go into C metric
```

Comment:
Can you explain why TEMP_FAHRENHEIT should point to "celsius"?

Reply:
Celsius is the [base unit for temperature](https://prometheus.io/docs/practices/naming/#base-units) in Prometheus and I believe HA converts temperature measurements to Celsius while exporting Prometheus metrics.

Reply:
These are not fixed units in Prometheus and changing these will be a massive breaking change for this integration.

I suggest on not doing that.

Reply:
Hi @frenck, thanks for chiming in. I would like to provide more context and understand your concerns better.

Prometheus documentation is very clear about [recommended base units](https://prometheus.io/docs/practices/naming/#base-units), and after a discussion amongst Prometheus integration contributors (home-assistant/architecture#391) we adopted hass-specific naming guidelines (#37149) that require following those documented base units. Most metrics exported by the Prometheus integration already follow the guidelines, except for a few older ones which I am fixing here.

Could you please clarify what you mean by a "massive" breaking change? This commit will rename a few metrics, which is clearly documented in the "Breaking change" section of the PR. Just as a point of reference, on my hass instance this resulted in 33 out of 720 metrics being renamed (less than 5% of them).

Are you opposed to any metric renaming, or do you only have specific concerns about changing the temperature unit suffix from `_c` to `_celsius`?

Since you are a hass maintaner who is in a position to either merge or reject this change, should I treat your comments as blocking? Or are you just providing personal recommendations and I should wait for another maintainer to make a decision on whether to accept this?

Thanks again for your feedback.

Reply:
Nothing in the linked architectural issue applies to this here.

> Prometheus documentation is very clear about recommended base units, 

Agreed. They are very clear on it being "recommended" and not mandatory.

> Could you please clarify what you mean by a "massive" breaking change? 

Anyone relying on the current namings anywhere in their usage/chain (of whatever they do with it) has to adjust to this change. The odd something having any of those renamed keys, is about 99% of the users of this integration.

> Since you are a hass maintaner who is in a position to either merge or reject this change, should I treat your comments as blocking?

🤷 I'm just leaving a comment with my opinion on that we should do this.

Reply:
> Nothing in the linked architectural issue applies to this here.

The home-assistant/architecture#391 issue is where we discussed the [metric naming plan](https://docs.google.com/document/d/1qkKObZXbZPVVVc6E4qHRpD5WNWcN0FrOeZFtL3HD7NE/edit#) that explicitly describes the renaming. Apologies for not linking to the doc directly.

> They are very clear on it being "recommended" and not mandatory. 

I am not sure it would be possible for Prometheus to really mandate any naming schema, since it's an open source piece of software that users can use however they want. But note that hass-specific prometheus naming guidelines introduced in #37149 actually use a stronger language ("Metric and label names should conform to Prometheus naming guidelines").

> Please note, I love fixing and improving things, but intentionally creating a breaking change for a naming recommendation is IMHO something we should never do; especially considering nothing is really broken either. A breaking change is not fun for anybody.

I agree that it will be an inconvenience, but I think it's important to keep in mind that this will not actually have any lasting impact (like data loss, etc).

I do appreciate your thoughts on this, and I would like to keep the discussion going and try to agree on a way to land this change in the most user-friendly way possible. However, you are sending a bit of a mixed message: in this PR you seem to be against changing names for a few metrics (if I understood you correctly), but in https://github.com/home-assistant/home-assistant.io/pull/17722 you suggested introducing a default namespace, which will effectively rename *all* metrics for existing users. This was originally proposed in https://github.com/home-assistant/architecture/issues/391#issuecomment-657249045, but I assumed it would be rejected by hass maintaners because of how disruptive it is.

What I would like to propose is actually to merge these two changes (1: introducing the default namespace; 2: bringing these few metrics in alignment with the naming guidelines) in a single PR, and include detailed instructions in the "Breaking change" section that will describe a way for existing users to keep the old names if they want. Specifically:
- users can specify an empty `namespace` to avoid the `hass_` metric name prefix.
- users can configure a set of `metric_relabel_configs` rules in Prometheus to counteract most of the naming changes I have in this PR now. A similar approach has been used to ease the migration in other large Prometheus metric renaming efforts like https://github.com/prometheus/node_exporter/issues/830.

As a result, I think we will greatly improve experience for new users, while providing a path for existing users to keep the old names if it's important to them. What do you think? If you are open to this approach, I am happy to spend more time developing and testing `metric_relabel_configs` and adjust this PR. Thanks!

Reply:
I think you misinterpreted my comment in the docs. I said it would be breaking, thus next suggested an additional example in the docs.

The external document linked in the architecture issue is not something part of the architectural repo and not a consideration for that matter.

Reply:
I agree that metric names should be normalized, ideally at the time of their introduction, not sure why that did not happen. 

How would we realize the recommendation to set a namespace for new users? This is really how it should be, because otherwise home-assistant pollutes lots of prefixes, making searches in Grafana less straight forward.

Ultimately these things impair the quality of the integration, lots of prometheus exporters all over the ecosystem fix up their metric names here and there, so that is really not uncommon. Also home-assistant regularly announces breaking changes, so there is already a possibility to announce these changes, so I'm hesitant to accept the resistance.

Reply:
I've discussed this with Martin, and we decided to move forward on this change. Not to be stuck in the past.

Reply:
Great news. Thank you, @frenck.

Next steps for me:
- bring this up to date with the `dev` branch.
- introduce the default namespace (`hass_`).
- expand the "breaking change" section providing more details on how users can keep the previous behaviour.

I will let you know when this is ready for another review.

Reply:
This is now ready for another review. I have extended the description of the breaking change, and [shared](https://gist.github.com/knyar/c3a58f8be92a2b7afadba5c16bbfc28f) some relabeling rules that can be used to keep the old metric names.